File: Day3/rest_api_project/routes/customers.js
// Customer routes
const express = require('express');
const router = express.Router();
const db = require('../db');

// GET all customers
router.get('/', async (req, res, next) => {
  try {
    const result = await db.query(
      'SELECT * FROM customers ORDER BY last_name, first_name'
    );
    res.json(result.rows);
  } catch (err) {
    next(err);
  }
});

// GET a specific customer by ID
router.get('/:id', async (req, res, next) => {
  try {
    const { id } = req.params;
    const result = await db.query(
      'SELECT * FROM customers WHERE customer_id = $1',
      [id]
    );
    
    if (result.rows.length === 0) {
      return res.status(404).json({ message: 'Customer not found' });
    }
    
    res.json(result.rows[0]);
  } catch (err) {
    next(err);
  }
});

// POST create a new customer
router.post('/', async (req, res, next) => {
  try {
    const { 
      first_name, 
      last_name, 
      email, 
      phone, 
      address, 
      city, 
      state, 
      zip_code 
    } = req.body;
    
    // Validate required fields
    if (!first_name || !last_name || !email) {
      return res.status(400).json({ 
        message: 'First name, last name, and email are required' 
      });
    }
    
    const result = await db.query(
      `INSERT INTO customers 
       (first_name, last_name, email, phone, address, city, state, zip_code) 
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8) 
       RETURNING *`,
      [first_name, last_name, email, phone, address, city, state, zip_code]
    );
    
    res.status(201).json(result.rows[0]);
  } catch (err) {
    // Check for duplicate email error
    if (err.code === '23505') {
      return res.status(400).json({ message: 'Email already exists' });
    }
    next(err);
  }
});

// PUT update a customer
router.put('/:id', async (req, res, next) => {
  try {
    const { id } = req.params;
    const { 
      first_name, 
      last_name, 
      email, 
      phone, 
      address, 
      city, 
      state, 
      zip_code 
    } = req.body;
    
    // Check if customer exists
    const checkResult = await db.query(
      'SELECT * FROM customers WHERE customer_id = $1',
      [id]
    );
    
    if (checkResult.rows.length === 0) {
      return res.status(404).json({ message: 'Customer not found' });
    }
    
    const result = await db.query(
      `UPDATE customers 
       SET first_name = $1, last_name = $2, email = $3, 
           phone = $4, address = $5, city = $6, 
           state = $7, zip_code = $8
       WHERE customer_id = $9
       RETURNING *`,
      [first_name, last_name, email, phone, address, city, state, zip_code, id]
    );
    
    res.json(result.rows[0]);
  } catch (err) {
    // Check for duplicate email error
    if (err.code === '23505') {
      return res.status(400).json({ message: 'Email already exists' });
    }
    next(err);
  }
});

// DELETE a customer
router.delete('/:id', async (req, res, next) => {
  try {
    const { id } = req.params;
    
    // Check if customer exists
    const checkResult = await db.query(
      'SELECT * FROM customers WHERE customer_id = $1',
      [id]
    );
    
    if (checkResult.rows.length === 0) {
      return res.status(404).json({ message: 'Customer not found' });
    }
    
    // Check if customer has orders
    const orderCheck = await db.query(
      'SELECT * FROM orders WHERE customer_id = $1 LIMIT 1',
      [id]
    );
    
    if (orderCheck.rows.length > 0) {
      return res.status(400).json({ 
        message: 'Cannot delete customer with existing orders' 
      });
    }
    
    await db.query('DELETE FROM customers WHERE customer_id = $1', [id]);
    
    res.json({ message: 'Customer deleted successfully' });
  } catch (err) {
    next(err);
  }
});

module.exports = router;

File: Day3/rest_api_project/python_version/routes/customers.py
"""Customer routes"""
from flask import Blueprint, request, jsonify
from db import query, query_one, execute
import psycopg2

customers_bp = Blueprint('customers', __name__, url_prefix='/api/customers')

# GET all customers
@customers_bp.route('', methods=['GET'])
def get_all_customers():
    try:
        results = query(
            'SELECT * FROM customers ORDER BY last_name, first_name'
        )
        return jsonify(results or [])
    except Exception as err:
        return jsonify({'error': 'Database error', 'message': str(err)}), 500

# GET a specific customer by ID
@customers_bp.route('/<int:id>', methods=['GET'])
def get_customer(id):
    try:
        result = query_one(
            'SELECT * FROM customers WHERE customer_id = %s',
            (id,)
        )
        
        if not result:
            return jsonify({'message': 'Customer not found'}), 404
        
        return jsonify(result)
    except Exception as err:
        return jsonify({'error': 'Database error', 'message': str(err)}), 500

# POST create a new customer
@customers_bp.route('', methods=['POST'])
def create_customer():
    try:
        data = request.get_json()
        
        first_name = data.get('first_name')
        last_name = data.get('last_name')
        email = data.get('email')
        phone = data.get('phone')
        address = data.get('address')
        city = data.get('city')
        state = data.get('state')
        zip_code = data.get('zip_code')
        
        # Validate required fields
        if not first_name or not last_name or not email:
            return jsonify({'message': 'First name, last name, and email are required'}), 400
        
        results = execute(
            '''INSERT INTO customers 
               (first_name, last_name, email, phone, address, city, state, zip_code) 
               VALUES (%s, %s, %s, %s, %s, %s, %s, %s) 
               RETURNING *''',
            (first_name, last_name, email, phone, address, city, state, zip_code)
        )
        
        return jsonify(results[0]), 201
    except psycopg2.IntegrityError as err:
        if 'unique constraint' in str(err):
            return jsonify({'message': 'Email already exists'}), 400
        return jsonify({'error': 'Database error', 'message': str(err)}), 500
    except Exception as err:
        return jsonify({'error': 'Database error', 'message': str(err)}), 500

# PUT update a customer
@customers_bp.route('/<int:id>', methods=['PUT'])
def update_customer(id):
    try:
        data = request.get_json()
        
        # Check if customer exists
        existing = query_one(
            'SELECT * FROM customers WHERE customer_id = %s',
            (id,)
        )
        
        if not existing:
            return jsonify({'message': 'Customer not found'}), 404
        
        first_name = data.get('first_name', existing['first_name'])
        last_name = data.get('last_name', existing['last_name'])
        email = data.get('email', existing['email'])
        phone = data.get('phone', existing['phone'])
        address = data.get('address', existing['address'])
        city = data.get('city', existing['city'])
        state = data.get('state', existing['state'])
        zip_code = data.get('zip_code', existing['zip_code'])
        
        results = execute(
            '''UPDATE customers 
               SET first_name = %s, last_name = %s, email = %s, 
                   phone = %s, address = %s, city = %s, 
                   state = %s, zip_code = %s
               WHERE customer_id = %s
               RETURNING *''',
            (first_name, last_name, email, phone, address, city, state, zip_code, id)
        )
        
        return jsonify(results[0])
    except psycopg2.IntegrityError as err:
        if 'unique constraint' in str(err):
            return jsonify({'message': 'Email already exists'}), 400
        return jsonify({'error': 'Database error', 'message': str(err)}), 500
    except Exception as err:
        return jsonify({'error': 'Database error', 'message': str(err)}), 500

# DELETE a customer
@customers_bp.route('/<int:id>', methods=['DELETE'])
def delete_customer(id):
    try:
        # Check if customer exists
        existing = query_one(
            'SELECT * FROM customers WHERE customer_id = %s',
            (id,)
        )
        
        if not existing:
            return jsonify({'message': 'Customer not found'}), 404
        
        # Check if customer has orders
        order_check = query_one(
            'SELECT * FROM orders WHERE customer_id = %s LIMIT 1',
            (id,)
        )
        
        if order_check:
            return jsonify({'message': 'Cannot delete customer with existing orders'}), 400
        
        execute('DELETE FROM customers WHERE customer_id = %s', (id,))
        
        return jsonify({'message': 'Customer deleted successfully'})
    except Exception as err:
        return jsonify({'error': 'Database error', 'message': str(err)}), 500